lib/axon/losses.ex

```diff
@@ -1024,4 +1024,170 @@ defmodule Axon.Losses do
       end
     )
   end
+
+  @doc """
+  WIP
+  y_pred should be log_softmax output of shape (B, T, D),
+  y_true should be of shape (B, S),
+  where
+  B - batch size,
+  T - sequence length,
+  D - dictionary size, including blank
+  S - maximum target length
+  y_true nas to be preencoded, its first and last characters should be
+  blank, each initial character should be separated with blanks,
+  double blanks mark the end of target sequence.
+  """
+  defn ctcloss(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, blank: 0, reduction: :none)
+    # eps = Nx.tensor([1.0e-7])
+    b_size = elem(Nx.shape(y_true), 0)
+    t_max = elem(Nx.shape(y_pred), 1) - 1
+    loss = Nx.broadcast(0.0, {b_size})
+
+    {loss, _, _, _} =
+      while {loss, b = 0, y_true, y_pred}, b < b_size do
+        # Get boundaries for available node paths.
+        st_lims = get_limits(y_true[b], t_max, opts[:blank])
+        # Iterate node tree backwards.
+        s_pred0 = iterate_tree(y_true[b], y_pred[b], st_lims, t_max)
+
+        {loss_b, _, _} =
+          while {loss_b = 0.0, s = st_lims[0][0], s_pred0}, s <= st_lims[0][1] do
+            {Nx.add(loss_b, Nx.exp(s_pred0[s])), s + 1, s_pred0}
+          end
+
+        loss_b = -Nx.log(loss_b)
+        {Nx.put_slice(loss, [b], Nx.reshape(loss_b, {1})), b + 1, y_true, y_pred}
+      end
+
+    transform(
+      {opts[:reduction], loss},
+      fn
+        # {:mean, loss} -> Nx.mean(loss)
+        {:sum, loss} -> Nx.sum(loss)
+        {:none, loss} -> loss
+      end
+    )
+  end
+
+  defn get_limits(y_true, t_max, blank) do
+    y_true_bb = Nx.concatenate([y_true, Nx.reshape(blank, {1})])
+    # Get `trimmed` tagret length.
+    {s_max, _} =
+      while {s = 0, y_true_bb}, y_true_bb[s] != y_true_bb[s + 1] do
+        {s + 1, y_true_bb}
+      end
```

Comment:
If I understand this right, we iterate over the tensor and select the first place where two adjacent values are the same? Off the top of my head, would this work? Assuming tokens are non-negative integers

```suggestion
s_max =
  y_true_bb
  |> Nx.window_reduce(-1, {2}, strides: [1], fn x, acc ->
    if acc == -1 do
      x # if I am seeing first value, just take it
    else
      x == acc # compare x2 and x1
    end
  end)
  |> Nx.multiply(Nx.iota({1, s})) # where s is the max targets?
  |> Nx.reduce_min() # take minimum index
```

The reason for avoiding while is that it is always slow

Reply:
I'll check this.

Reply:
I have some issues with `window_reduce`.
This, for example, gives me a CompileError:
```elixir
defn test() do
  Nx.window_reduce(Nx.tensor([1, 2, 3]),
    0, {2}, [padding: :same],
    fn x, acc -> x + acc end
  )
end
```

Reply:
What's the error?

Reply:
<details>
<summary>Error log</summary>

```shell
== Compilation error in file lib/axon_ctc.ex ==
** (MatchError) no match of right hand side value: 1
    (nx 0.1.0-dev) lib/nx/binary_backend.ex:623: Nx.BinaryBackend.element_wise_bin_op/4
    (nx 0.1.0-dev) lib/nx/defn/evaluator.ex:147: Nx.Defn.Evaluator.eval_apply/4
    (nx 0.1.0-dev) lib/nx/defn/evaluator.ex:63: Nx.Defn.Evaluator.eval/3
    (nx 0.1.0-dev) lib/nx/defn/evaluator.ex:83: anonymous fn/4 in Nx.Defn.Evaluator.eval_apply/4
    (nx 0.1.0-dev) lib/nx/binary_backend.ex:1353: Nx.BinaryBackend."-window_reduce/6-_@5/2-2-"/4
    (nx 0.1.0-dev) lib/nx/binary_backend.ex:1351: anonymous fn/9 in Nx.BinaryBackend.window_reduce/6
    (elixir 1.12.2) lib/enum.ex:2385: Enum."-reduce/3-lists^foldl/2-0-"/3
    (nx 0.1.0-dev) lib/nx/binary_backend.ex:1345: Nx.BinaryBackend.window_reduce/6
```
</details>

Reply:
This seems to be a bug, can you extract a minimal example and post an issue in the Nx repo?

Reply:
I suggest we leave it as it is for now, and I'll come back to this when `window_reduce` bug will be settled.

Reply:
Done.